### DISCUSSION TITLE
Python and Transcrypt Version Update

### DISCUSSION THREAD

**Discussion by JennaSys** [rtp-book/project#1] [Errata]
At the time the _React to Python_ book was written, Transcrypt only worked with Python version 3.7. However a new version of Transcrypt was recently released that works with Python 3.9. This is now the default version on [PyPI](https://pypi.org/project/Transcrypt/), and if no version is specified when you use PIP to install it, you will get the newer 3.9 version.  If you are are using Python 3.7 as the book describes, this will likely result in compilation errors.  So you now have two options:
1. Continue to use Python 3.7 in which case you will have to specify the version of Transcrypt to use when you install it:
`pip install transcrypt==3.7.16`
2. Use Python 3.9 instead of Python 3.7 when running the examples and just install the latest version of Transcrypt:
`pip install transcrypt`

While not all of the examples from the book have been fully tested with Python 3.9, they should be compatible with it.  Other than installing the correct version of Transcrypt based on the Python version you are using, no other changes should be needed.

**Comment by dvshmkr** [comment]
I bought your book a couple years ago and  I'm just now finally digging into it. I'm working on the first 'Hello World' example.  I running on opensuse leap 15.4 which is running Python 3.6.15. is that going to work with your examples? I did back the trancrypt version down to 3.6.101 but I'm still getting errors:

```
(venv) dave@web5:~/projects/python> transcrypt hello        

Transcrypt (TM) Python to JavaScript Small Sane Subset Transpiler Version 3.6.101
Copyright (C) Geatec Engineering. License: Apache 2.0


list index out of rangeTraceback (most recent call last):
  File "/home/dave/projects/python/venv/lib64/python3.6/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 167, in __init__
    moduleMetadata = ModuleMetadata (self, self.mainModuleName)
  File "/home/dave/projects/python/venv/lib64/python3.6/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 68, in __init__
    '\n\tCan\'t find any of:\n\t\t{}\n'.format ('\n\t\t'. join (searchedModulePaths))
org.transcrypt.utils.Error: <unprintable Error object>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dave/projects/python/venv/lib64/python3.6/site-packages/transcrypt/__main__.py", line 159, in main
    compiler.Program (transpilationDirs, modulesDir, __symbols__)
  File "/home/dave/projects/python/venv/lib64/python3.6/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 172, in __init__
    message = '\n\t{}'.format (str (exception))
  File "/home/dave/projects/python/venv/lib64/python3.6/site-packages/transcrypt/modules/org/transcrypt/utils.py", line 149, in __str__
    result += '\n\tFile \'{}\', line {}, namely:'.format (program.importStack [-1][0] .sourcePath, self.lineNr)
IndexError: list index out of range


Aborted
```

Any clues -

**Comment by dvshmkr** [comment]
I decided to try on Rocky Linux since OpenSuse is stuck on V3.6 of Python. Hey, I got it working! I had to use the --nomin option though. I saw the note in the book for Windows users but does the same apply to linux if Java is not installed?

Also, I'd really prefer to work on Opensuse. Can you advise which verison of Transcrypt I should use with Python 3.6.15? Is there a way for me to make that work?

Thanks

**Reply by JennaSys** [reply]
The book uses Python 3.7 since that is the version that Transcrypt used at the time. The latest version Transcrypt now uses Python 3.9. The "[Python and Transcrypt Version Update](https://github.com/rtp-book/project/discussions/1)" post covers the details on how to use either one.

So that said, if you don't have Python 3.7 or Python 3.9 on your system you will ahve to install one of those in order to use Transcrypt. You can have multiple versoin of Python on the same system. However, when you install another version of Python, do NOT replace the version of Python that came with your OS. Leave that one in place and install the new version of Python in addition to it.

Then, when you create the virtual environment using `venv`, be specific about the version of Python that you run. Something like this:
`python3.9 -m venv venv`
When the virtual environment is activated, just running `python` will use the correct version.

**Comment by dvshmkr** [comment]
**Problem**
I've made it to Chapter 10 covering the Parcel package bundler. I read through your update info regarding parcel v2 but I was hoping to complete the excercises in the book as written using the previous version of Parcel. I did get a number of deprecation warning when I installed Parcel with _npx install parcel-bundler --save-dev_. 

When I build the files as directed on page 52, no errors are retuned. The app comes up in the browser but there is no content in the 'root' div. The developer console shows `Uncaught SyntaxError: Unexpected identifier 'pyreact4' (at react_hello4.57fd4402.py:1:6)`. It's as though the react_hello4.py can't find pyreact4.py but it is in the same folder along with the html file.

**Question**
Is this due to the using the old version of Parcel meaning I'll have to follow the  Parcel v2 update procedures for the remainder of the book?

**Reply by JennaSys** [reply]
The old version of Parcel should still work, and there may be some benefit to keep using that with the book to avoid any confusion while you are going through it. That said, the new Parcel V2 version is definitely better, and if you do any projects beyond the book I'd recommend moving to that one for anything else.

If you look in the \_\_target\_\_ folder that Transcrypt produces, do the pyreact4.py and pyreact4.js files exist?
When you run the parcel command, does the log output show that pyreact4.js is being built?

**Reply by dvshmkr** [reply]
Well, my build is not producing a` __target__` folder. I have the dist folder containing _react_hello4.html_ and _react_hello4.#####.py_.

Output from the build:
```
(rtp) [dave@localhost ch10]$ npx parcel --log-level 4 --no-cache react_hello4.html
[9:25:39 PM]: Server running at http://localhost:1234 
[9:25:39 PM]: Building...
[9:25:40 PM]: Building react_hello4.html...
[9:25:40 PM]: Building react_hello4.py...
[9:25:41 PM]: Built react_hello4.py...
[9:25:41 PM]: Built react_hello4.html...
[9:25:41 PM]: Producing bundles...
[9:25:41 PM]: Packaging...
[9:25:41 PM]: ✨  Built in 1.68s.
```
The little star emoji on the last line is just a rectangle/non-printable char on my console. Not sure if that might be a clue. I did make the edit to _asset.js_ as directed on page 49. I notice that _pyreact4.py_ is not listed in the build? I seems like the import on line 1 of _react_hello4.py_ is not being recognized. I'll bring my laptop to the next meeting and we can look it over.

Thanks again for your assistance.

**Reply by JennaSys** [reply]
Hmmm there definitely should be a \_\_target\_\_ folder being created. That's what gets fed into the Parcel bundler.
